Add cross platform support for kitchen-pester

gaelcolas opened the pull request:
I need kitchen-pester to work on several OSes (ubuntu, Suse, RHEL and others), and I think it could do with some cleanup and refactor in some areas.

I'm opening this draft PR so it's easier to provide visibility in the directions I'm thinking, and so I can add code comments to discuss change before actually implementing them.

In its current implementation, Test-Kitchen has some challenges to execute PowerShell script on non-windows platform (oob), because the transport is ssh, and replacing the default shell with something like `chsh` is not trivial in the lifecycle hooks, and depends a bit of the package manager used (snap, apt, yum...).

What should work across Linux systems is using `#!/usr/bin/env pwsh` on invoking a script to be deterministic on invoking something in pwsh, so I'll try to experiment in this direction, without touching too much the Windows part.

Any suggestion welcome, you can also reach out to me on the PS slack to discuss.

gaelcolas commented:
First question/suggestion, move PowerShell code to `PesterUtil.psm1` or adjacent module.
Having here-strings in the ruby code is less than ideal for development.
The ruby code should have only the minimum: Import-module and call the command from that module(s)

gaelcolas commented:
Last edit suggests how we could try to write each command to file with correct shebang in the `really_wrap_shell_code(code)` function, and get that file executed when not running on windows.
This pose the questions on how does it currently work on Windows when you want to use `pwsh` or `pwsh-preview`.
I also need to check how/if the Pester output is streamed back to the local machine from PS for display.

gaelcolas commented:
Added a suggestion to copy a configured list of folders so that the local machine can handle the management of modules pushed to the SUT (increased time in pushing the files over transport, but more flexibility).

gaelcolas commented:
We should get rid of the whole PSGet "feature", it's from an other time...
Maybe we can bootstrap PowerShellGet instead (v2 for now, find an option for v3?)

Also when installing from GH we should try Expand-Archive first (PS5+) and fall back to using COM object for PS4-.

vexx32 commented:
@gaelcolas pretty much entirely in agreement with all your suggestions so far, just haven't had the time / dire need to get it done yet. Thanks for taking the initiative!! 💖 😊 🌸

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -211,7 +221,11 @@ def install_command_script
               }
               else {
-                  if (-not (Test-Module -Name PsGet)){
+                  if (-not (Test-Module -Name PsGet)){ # We should get rid of this. 
+                      # installing PSGet from someone we don't know's github raw is scary bad.
+                      # We don't want PSGet but maybe PowerShellget instead.
+                      # Maybe install from nupkg directly from a nuget feed (Odata), doing the nupkg unzip ourselves?
+                      # and we need PowerShellGet, v2 or v3
```
Yep, agreed. I left PsGet in for the time being because I wasn't sure if PowerShellGet could be used downlevel, but as far as I'm aware it's usable down to v2 IIRC, provided we can bootstrap it.

gaelcolas replied:
And I guess you need it to support PSv2? 🙄

vexx32 replied:
I wish I could say no, I don't like dealing with PSv2, but... yep. 😔

gaelcolas replied:
as discussed, tested with PS4 only in 2012r2...

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -152,6 +157,10 @@ def run_command_script
       end
 
       def really_wrap_shell_code(code)
+        # hypothesis: if OS not windows (can we detect or assume from transport)
+        # write the wrapped shell code to file with the pwsh(-preview) shebang
+        # and execute the file
+        # leave as is for windows (but double check we can use pwsh & pwsh-preview too)
```
I also don't know for sure, but there is very likely a way to determine from the transport and/or the instance itself what kind of OS it is, yeah. 

For non-Windows if we plan to invoke the files directly via shebang we'll also need to add some code to set the executable bit of the file(s).

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -227,10 +241,13 @@ def install_command_script
                       }
 
-                      Invoke-Expression -Command $webClient.DownloadString('http://bit.ly/GetPsGet')
+                      Invoke-Expression -Command $webClient.DownloadString('http://bit.ly/GetPsGet') # this resolves to https://raw.githubusercontent.com/chaliy/psget/master/GetPsGet.ps1
+                      # then in turns installs https://github.com/psget/psget/raw/master/PsGet/PsGet.psm1
+                      # we should get rid of this relic
                   }
 
                   try {
+                    # We should change the below to PowerShellGet (I assume)
```
Yep, pretty much 🙂

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -266,6 +283,7 @@ def install_command_script
                           Write-Host "Downloaded Pester.zip"
                       }
 
+                      # Try Expand-Archive first, only fall back to COM... (PS < 5.1)
```
Yeah, agreed. I'd just do a `Get-Command Expand-Archive` rather than a version check, just in case folks have really weird things happening on their instances. 😂

smurawski commented:
I'm fine if we major version bump and drop PS2 support and move to PowerShell 6+.

vexx32 commented:
I'd prefer to keep 5.1 as the minimum, personally, but if we're doing a major version bump it's probably fine to go to 6+.

gaelcolas commented:
I'm usually much more aggressive towards legacy, but given that I see Test-Kitchen as a key tool for folks to automate away their old systems, I'll try to stay compatible with WMF 4.0 as it corresponds to 2012r2 oob.
Although, duly noted, next major update we'll probably get rid of that too.

Given the changes I'm making, we'll check towards the end to see if we want to bump the major version anyway, or if we assume that a major version of 0 means we can do what we want.

gaelcolas commented:
I've hit a bug with test-kitchen I think.
When I create the sandbox, and do a `kitchen verify` the first time it does not preserve the folder structure for some reasons. 
For instance, I try to have the `PesterUtil.psm1` under `$Env:Temp\verifier\modules\PesterUtil\`, but on the first run it will do:
```
C:\USERS\AZURE\APPDATA\LOCAL\TEMP
+ kitchen
|       suite1_script.ps1
|       
+ verifier
|   +PesterUtil
|   |       PesterUtil.psm1
|   |       
+   +suites
|           default.tests.ps1
|
+ winrm-upload
```

But the second time and any subsequent try:

```
C:\USERS\AZURE\APPDATA\LOCAL\TEMP
|   
+ kitchen
|       suite1_script.ps1
|       
+ verifier
|   |   
|   + modules
|   |   + PesterUtil
|   |           PesterUtil.psm1
|   |
|   + PesterUtil
|   |       PesterUtil.psm1
|   |       
|   + suites
|           default.tests.ps1
|           result.xml
|
+ winrm-upload
```

This issue was not apparent before in kitchen-pester because the install steps did not rely on `PesterUtil.psm1`.
As far as I can tell, it's done by the same code [in the base verifier](https://github.com/test-kitchen/test-kitchen/blob/990f89768418c5582cc52c1ef40651a2d54239da/lib/kitchen/verifier/base.rb#L74).

I'm a bit at lost as to how to troubleshoot that further, so I'll consider some dirty workarounds. :(
FTR I'm on `Test-Kitchen v2.5.1` and installed the latest ChefDK 4.8.23.

@smurawski, any suggestion welcome.

gaelcolas commented:
Ooohhhh, I found that it only happens if `PesterUtil` is the only folder in the `modules` folder.
So I worked around that by creating another folder (and a file so that it's not ignored at gem compilation time) next to my `PesterUtil` folder and it just works.
![image](https://user-images.githubusercontent.com/8962101/83972798-22e52180-a8da-11ea-8a6d-4e0b689b3856.png)
 
It's not pretty, but at least it solves the issue without making weird code to handle that inconsistency.

I can move on...

gaelcolas commented:
there's more to it... :(
The problem had moved before coming back to that...

gaelcolas commented:
Getting closer, after a break...
It's now working fine on first and subsequent runs.

I've added a `copy_folders` feature so we can copy local modules to remote.
For now they are moved those under `modules` folder, so that the PSModulePath can discover them.
The bootstrap modules are only installed to `$PSHome/modules` if not present. Default with nothing assuming the image you use is ok. You can force the install.

Pester install as we discussed is a splat to install module.

Register repository is not implemented yet, maybe the next step. Probably just splatting to `Register-PSRepository` if the repository name does not exist, and splat to `Set-PSRepository` if it does.

```
verifier:
  name: pester
  test_folder: tests/Integration
  bootstrap:
    repository_url: https://www.powershellgallery.com/api/v2
    modules:
    - Name: PowerShellGet
      Version: 2.2.4.1
      Force: true
    - Name: PackageManagement
      Version: 1.4.7
  pester_install:
    Name: pester
    RequiredVersion: 4.10.1
    Force: true
    AllowPrerelease: true
  install_modules:
    - Name: Datum
      Force: true
    - Name: Datum.ProtectedData
      MaximumVersion: 0.1.1
      AllowPrerelease: false
      Force: true
  copy_folders:
    - output/RequiredModules/Pester
  register_repository: []
    # - Name: my_repo
    #   Url: <uyrl>
```
/cc @vexx32

gaelcolas commented:
It's not working on linux yet btw, I'm experimenting some ways, not very successfully so far :/

gaelcolas commented:
Today is a good day:
![image](https://user-images.githubusercontent.com/8962101/84597505-4240f900-ae5c-11ea-8a89-34f0a45fd6ab.png)

Now Kitchen-pester works cross platform and many features have been added/updated.
It's only missing the Set/Register PSRepository feature, which should not take too long.

I'd appreciate you have a look at this to tell me what you think before I start the formal review process.
Also, I might need some help to understand what's going on in the CI pipeline and why it fails.

gaelcolas commented:
Marking as Ready for review as I've implemented about everything I had in mind and more.
I only have a bug on downloading the Pester xml result on Linux, I get some SCP error I'll have to check why.
I've added the uninstall of built-in modules (PackageManagement, PSGet, Pester) because those early versions cause all kind of problems.
The removal happens by default (respective version of 1.0.0.1, 1.0.0.1, 3.4.0), but can be disabled.

There's a lot of documentation that will need updating, but for now I'm happy with most of the code, and would like another pair of eyes.

My test config looks like this:
```yaml
---
driver:
  name: azurerm
  subscription_id: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
  location: 'westus2'
  machine_size: 'Standard_D2s_v3'

provisioner:
  name: shell
  script:    'tests/Integration/shell_suite1/suite1_script.ps1' # Optional - default: bootstrap.sh/bootstrap.ps1

verifier:
  name: pester
  test_folder: tests/Integration
  bootstrap:
    repository_url: https://www.powershellgallery.com/api/v2
    modules:
    - Name: PackageManagement
      Version: 1.4.7
    - Name: PowerShellGet
      Version: 2.2.4.1
  pester_install:
    RequiredVersion: 4.10.1
    Force: true
    SkipPublisherCheck: true
  install_modules:
    # - Name: Datum
    #   Force: true
    # - Name: Datum.ProtectedData
    #   MaximumVersion: 0.1.1
    #   AllowPrerelease: false
    #   Force: true
  copy_folders:
    # - output/RequiredModules/Pester
  register_repository: []
    # - Name: my_repo
    #   Url: <uyrl>
  # downloads:

platforms:
  - name: ubuntu-18.04
    driver:
      image_urn: Canonical:UbuntuServer:18.04-LTS:latest
    lifecycle:
      post_create:
      - remote: sudo snap install powershell --classic
    verifier:
      downloads: # That does not work on linux, need to fix
      bootstrap:
        modules: # override because no need to bootstrap PS7

  - name: win2019
    driver:
      image_urn: MicrosoftWindowsServer:WindowsServer:2019-Datacenter:latest

  - name: win2012r2
    driver:
      image_urn: MicrosoftWindowsServer:WindowsServer:2012-R2-Datacenter:latest

suites:
  - name: default
```

Technically most of the config should go and just use the `copy_folders` feature to copy the module saved locally to the SUT. I need to test that xplat just to be sure.

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -93,8 +97,38 @@ def create_sandbox
+
+          if ($#{config[:remove_inbox_pester]}) {
+            $modulesToRemove += Get-module -ListAvailable -FullyQualifiedName @{ModuleName = 'Pester'; RequiredVersion = '3.4.0'}
+          }
```
We'll do thorough review later of course but saw this and figured I'd suggest this for now anyway 💖 

```suggestion
          $modulesToRemove = @(
              if ($#{config[:remove_inbox_packagemanagement]}) {
                  Get-Module -ListAvailable -FullyQualifiedName @{ModuleName = 'PackageManagement'; RequiredVersion = '1.0.0.1'}
              }

              if ($#{config[:remove_inbox_psget]}) {
                  Get-Module -ListAvailable -FullyQualifiedName @{ModuleName = 'PowerShellGet'; RequiredVersion = '1.0.0.1'}
              }

              if ($#{config[:remove_inbox_pester]}) {
                  Get-Module -ListAvailable -FullyQualifiedName @{ModuleName = 'Pester'; RequiredVersion = '3.4.0'}
              }
          )
```

gaelcolas replied:
Yeah completely agree.

vexx32 commented:
@gaelcolas awesome, thanks! I'll make some time tomorrow to have a good look at it 😊 💖

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -369,33 +422,81 @@ def prepare_helpers
+      def ps_hash(obj, depth = 0)
+        if [true, false].include? obj
+          %{$#{obj}} # Return $true or $false when value is a bool
+        elsif obj.is_a?(Hash)
+          obj.map do |k, v|
+            # Format "Key = Value" enabling recursion
+            %{#{pad(depth + 2)}#{ps_hash(k)} = #{ps_hash(v, depth + 2)}}
+          end
+            .join(";\n")  # append ;\n to the key/value definitions
+            .insert(0, "@{\n") # prepend @{\n 
+            .insert(-1, "\n#{pad(depth)}}") # append }\n
```
```suggestion
        elsif obj.is_a?(Hash)
          obj.map do |k, v|
            # Format "Key = Value" enabling recursion
            %{#{pad(depth + 2)}#{ps_hash(k)} = #{ps_hash(v, depth + 2)}}
          end
            .join("\n")  # append \n to the key/value definitions
            .insert(0, "@{\n") # prepend @{\n 
            .insert(-1, "\n#{pad(depth)}}\n") # append \n}\n
```

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,143 @@
+[Net.ServicePointManager]::SecurityProtocol = [Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12
+
+function Install-ModuleFromNuget {
+    param (
+        $Module,
+        $GalleryUrl         = 'https://www.powershellgallery.com/api/v2'
+    )
```
```suggestion
    param(
        [string]
        $Module,
        
        [uri]
        $GalleryUrl = 'https://www.powershellgallery.com/api/v2'
    )
```

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,143 @@
+
+    if ((Test-Path $ModuleFolder) -and ($PSVersionTable.PSVersion.Major -lt 5 -or $module.Force)) {
+        # Check if available version is correct
+        $ModuleManifest = (Join-Path -Path $ModuleFolder -ChildPath "$($Module.Name).psd1")
```
```suggestion
        $ModuleManifest = Join-Path -Path $ModuleFolder -ChildPath "$($Module.Name).psd1"
```

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,143 @@
+    if ((Test-Path $ModuleFolder) -and ($PSVersionTable.PSVersion.Major -lt 5 -or $module.Force)) {
+        # Check if available version is correct
+        $ModuleManifest = (Join-Path -Path $ModuleFolder -ChildPath "$($Module.Name).psd1")
+        if ((Test-Path -Path $ModuleManifest) -and -not $Module.Force) {
+            # Import-PowerShellDataFile only exists since 5.1
+            $ManifestInfo = Import-LocalizedData -BaseDirectory (Split-Path -Parent -Path $ModuleManifest) -FileName $Module.Name
+            $ModuleVersionNoPreRelease = $Module.Version -replace '-.*',''
+            # Compare the version in manifest with version required without Pre-release
+            if ($ManifestInfo.ModuleVersion -eq $ModuleVersionNoPreRelease) {
+                Write-Host "Module $($Module.Name) already installed, skipping."
+                return
+            }
+            else {
+                Write-Host "Module $($Module.Name) found with version '$($ManifestInfo.ModuleVersion)', expecting '$ModuleVersionNoPreRelease'."
+            }
+        }
```
```suggestion

        if ((Test-Path -Path $ModuleManifest) -and -not $Module.Force) {
            # Import-PowerShellDataFile only exists since 5.1
            $ManifestInfo = Import-LocalizedData -BaseDirectory (Split-Path -Parent -Path $ModuleManifest) -FileName $Module.Name
            $ModuleVersionNoPreRelease = $Module.Version -replace '-.*'

            # Compare the version in manifest with version required without Pre-release
            if ($ManifestInfo.ModuleVersion -eq $ModuleVersionNoPreRelease) {
                Write-Host "Module $($Module.Name) already installed, skipping."
                return
            }
            else {
                Write-Host "Module $($Module.Name) found with version '$($ManifestInfo.ModuleVersion)', expecting '$ModuleVersionNoPreRelease'."
            }
        }
```

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,143 @@
+    }
+    elseif ($PSVersionTable.PSVersion.Major -gt 5) {
+        # skip if the version already exists or if force is enabled
+        $ModuleVersionNoPreRelease = $Module.Version -replace '-.*',''
```
```suggestion
        $ModuleVersionNoPreRelease = $Module.Version -replace '-.*$'
```

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    $urlSuffix = "/package/$($Module.Name)/$($Module.Version)"
+    $nupkgUrl = $GalleryUrl.TrimEnd('/') + '/' + $urlSuffix.Trim('/')
```
```suggestion
    $nupkgUrl = New-Object -TypeName System.Uri -ArgumentList $GalleryUrl, $urlSuffix
```

URI constructor is a little less messy than manual string parsing. 🙂

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,143 @@
+            Register-PSRepository @Repository
+        }
+    }
+    else {
+        Write-Host "Cannot Set PS Repository, command Set or Register for PSRepository or PackageSource not found."
+    }
```
Should we throw an error in this case?

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -33,8 +36,20 @@ class Pester < Kitchen::Verifier::Base
 
       default_config :restart_winrm, false
       default_config :test_folder
+      default_config :remove_inbox_packagemanagement, true
+      default_config :remove_inbox_psget, true
```
the two are less interdependent in some versions, but for the built in you are right. They should be removed together.
do you have a goof variable name? :)

vexx32 replied:
🤷 Not at the moment. I'm half inclined to just say use `remove_builtin_powershellget` or something and have that cover both, I'm not sure.

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -33,8 +36,20 @@ class Pester < Kitchen::Verifier::Base
+      default_config :remove_inbox_psget, true
+      default_config :remove_inbox_pester, true
       default_config :use_local_pester_module, false
+      default_config :bootstrap, {
+        :repository_url => "https://www.powershellgallery.com/api/v2",
+        :modules => []
+      }
```
hmmm, I don't think so...
If we agree that THE powershell version is 7.x, then it's not necessary.
You only need PowerShellGet & PackageManagement IF your system is not setup properly (aka defaults for PS4 & PS 5.1).

Not doing that means we have to override the bootstrap when using on PS 7 so it does not bootstrap a version.

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -33,8 +36,20 @@ class Pester < Kitchen::Verifier::Base
+        :modules => []
+      }
+      default_config :register_repository, []
+      default_config :pester_install, {"Name" => "Pester", "SkipPublisherCheck" => true, "Force" => true, "ErrorAction" => "Stop" }
```
My initial idea was to let someone use something else such as Gherkin, but it's pointless because the invocation is Pester specific. Will change.

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -33,8 +36,20 @@ class Pester < Kitchen::Verifier::Base
+      default_config :install_modules, []
       default_config :downloads, ["./PesterTestResults.xml"] => "./testresults"
+      default_config :copy_folders, []
+      default_config :sudo, true
```
Yeah this is only to run on non-windows.
On windows  everything runs through winrm, in others it uses SSH as transport.
Because of this, ssh needs to call pwsh, which can be tricky permission-wise without customizing the OS (depending how pwsh has been installed).
so everytime Kitchen-Pester runs a PS command, this will run as `sudo pwsh -f...` if true.

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -75,11 +97,40 @@ def create_sandbox
-        return if local_suite_files.empty?
-        return if config[:use_local_pester_module]
+        # the sandbox has not yet been copied to the SUT.
+        install_command_string = <<-CMD
```
yeah, it's a ruby here string.
You are right, I should be consistent. Will fix.

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -151,145 +204,156 @@ def run_command_script
+        bootstrap = config[:bootstrap]
+        # if the repository url is set, use that as parameter to Install-ModuleFromNuget. Default is the PSGallery url
+        repository_url = bootstrap[:repository_url]
+        gallery_url_param = if repository_url
+          "-GalleryUrl '#{repository_url}'"
+        else
+          ""
+        end
```
It would break the powershell, but similarly as typing the wrong URL...
The ruby side should be fine.
Not sure what would be better.... %{} maybe...?

vexx32 replied:
Perhaps build it into a hash for splatting? 🤔

gaelcolas replied:
feels overkill and less visible if we create a ruby hash before converting it to PS then splat.
Maybe even more fragile...

gaelcolas replied:
I haven't changed this for now, as I don't see what would be really better...

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -151,145 +204,156 @@ def run_command_script
+            pwsh_cmd = "pwsh"
+          end
+
+          myCommand = <<-EOH
+          echo "Running as '$(whoami)'"
+          cat << 'EOF' > current.ps1
+          #!/usr/bin/env pwsh
+          #{Util.outdent!(use_local_powershell_modules(code))}
+          EOF
+          mkdir -p foo #{config[:root_path]}/modules
+          #{pwsh_cmd} -f current.ps1
+          EOH
```
I should repl;ace the EOH with BASH,
That's just a bash script to create a file from the generated heredoc, then invoking pwsh -f.
I'll see what I can do.

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -151,145 +204,156 @@ def run_command_script
+            $null = New-Item -Path $PSModPathToPrepend -Force -ItemType Directory
+          }
+          
+          if ($Env:PSModulePath.Split([io.path]::PathSeparator) -notcontains $PSModPathToPrepend) {
+            $env:PSModulePath   = @($PSModPathToPrepend, $env:PSModulePath) -Join [io.path]::PathSeparator
+          }
```
That wouldn't work if there's already a subpath of the path I want to add in PSModulePath.
`'c:\src\sub\path;c:\src\sub' -notmatch [regex]::Escape('c:\src')`

gaelcolas commented:
back on this after a long break (sorry).

gaelcolas commented on README.md:
```diff
@@ -28,6 +30,8 @@ This can be a top-level declaration, a per-node declaration, or a per-suite decl
 
 * `restart_winrm` - boolean, default is false.  This is primarily to support powershell v2 scenarios.  If Pester is not being found, enable this option.
 * `test_folder` - string, default is nil.  `test-folder` allows you to specify a custom path (the default is ./test/integration/) for your integration tests.  This can be an absolute path or relative to the root of the folder kitchen is running from.  This path must exist.
+* `use_local_pester_module` - boolean, default is false.
+* `downloads`- map[string, string], defaults to `["./PesterTestResults.xml"] => "./testresults"`. Files to download from SUT to local system, used to download the pester results locally. The key is the remote file, value the local file it should be saved as.
```
yeah, will do last when everything else is ok as we risk changing params & names.

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -151,145 +204,156 @@ def run_command_script
+        $InstallPesterParams = #{ps_hash(pester_install_params)}
+        if (!$InstallPesterParams.ContainsKey('Name')) {
+          $InstallPesterParams['Name'] = 'Pester'
+        }
```
hard coded.

gaelcolas commented:
I'm reverting to my `$nupkgUrl = $GalleryUrl.TrimEnd('/') + '/' + $urlSuffix.Trim('/')` because I remember why it's "best"...
The System.Uri constructor does not work when the uri already has some relative path (`/api/v2`) so adding the suffix (`$packageName/$PackageVersion`) fails on the constructor...

gaelcolas commented:
Working on https://github.com/gaelcolas/kitchenSamples/blob/master/Demo1/kitchen.yml for proper testing.
Windows 2019 and 2012r2 works, Ubuntu not anymore for some reasons.

I might add some better default logic so that it's a bit easier on windows, and I need to try pwsh on win.

gaelcolas commented:
I think that's working well now, and is as simple as can be for a basic setup (assuming a working PowerShellGet).
The documentation needs to be done, but I want to make sure we're happy with functionality and variable names first.

```yaml
verifier:
  name: pester
  copy_folders:
    - MyModule
  pester_install:
    MaximumVersion: '4.99.999'
```

When the system does not have a decent version of PowerShellGet & PackageManagement (i.e. defaults on Windows), it needs to be bootstrapped (instead of using the `-Forcebootstrap` which is temperamental).

For this, you can add the bootstrap config as below:
```yaml
  bootstrap:
    repository_url: "https://www.powershellgallery.com/api/v2"
    modules:
      - PackageManagement
      - Name: PowerShellGet
        Version: '2.2.4.1'
```

This will download the nupkg package for the module (using the nuget url without the need for nuget binary).

My testing example is this: https://github.com/gaelcolas/kitchenSamples/tree/master/Demo1

Let me know any feedback. /cc @vexx32 @smurawski

gaelcolas commented:
Forgot to mention that I'm not sure whether it works with Pester 5, although it might as I think @nohwnd has added command & parameter aliases...?
The Invocation of pester could probably be re-worked, but I'd rather do that in another iteration.

glennsarti commented on build.windows.ps1:
```diff
@@ -0,0 +1,10 @@
+#! 
+# https://www.sitepoint.com/creating-your-first-gem/
+chef gem build $PSScriptRoot/kitchen-pester.gemspec
```
Do you need to call `chef` here?  `gem build ...` will build the gem.

vexx32 replied:
Hmm, yeah. Should do. Assuming `gem` is on the path. Maybe safer to assume `chef` will be on the path, and it's essentially guaranteed to come with a working `gem` package?

gaelcolas replied:
Yeah, given the context (TK), I'd rather stick with chef.
Also, this is a dirty helper that is convenient for testing, but far from critical.
I think it's mainly of use for noob like me trying to understand how to compile and install the gem.

glennsarti commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -148,148 +199,173 @@ def run_command_script
+        bootstrap = config[:bootstrap]
+        # if the repository url is set, use that as parameter to Install-ModuleFromNuget. Default is the PSGallery url
+        repository_url = bootstrap[:repository_url]
+        if repository_url
```
nit : Can use a ternary here

`gallery_url_param = repository_url ? "-GalleryUrl '#{repository_url}'" : ""`

glennsarti commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -148,148 +199,173 @@ def run_command_script
+            # so we can splat that variable to install module
+            <<-PS1
+              $#{module_name} = #{ps_hash(powershell_module)}
+              Write-host -noNewline 'Instaling #{module_name}'
```
nit: typo

vexx32 replied:
also: PascalCase for PS commands and parameters please! 🙂

gaelcolas replied:
I tend to use camelCase for local variables and PascalCase for parameters or class attributes, but I agree that without PSSA rules, I'm not always consistent.

gaelcolas replied:
Although here it's all ruby variables hence he snake_case... (`#{ }` interpolations)

glennsarti commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -148,148 +199,173 @@ def run_command_script
 
       def really_wrap_shell_code(code)
-        wrap_shell_code(Util.outdent!(use_local_powershell_modules(code)))
+        if windows_os?
```
instead of doing an `if` with an indent, use a guard clause.

```
return wrap_shell_code(Util.outdent!(use_local_powershell_modules(code))) if windows_os?
```

or perhaps abstract it further: e.g.

```
def really_wrap_shell_code(code)
  windows_os? ? really_wrap_windows_shell_code(code) : really_wrap_posix_shell_code(code)
end

def really_wrap_windows_shell_code(code)
  ...
end

def really_wrap_posix_shell_code(code)
  ...
end
```

vexx32 replied:
Hmm. I think it'll be clearer to abstract it one layer.

vexx32 replied:
Additional Q: is it worth having a branch here for MacOS, or is it safe enough to treat it similarly to Linux?

gaelcolas replied:
I think it should be the same for MacOS, it's mainly about shebang, sudo and wrapping for pwsh invocation...

glennsarti commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -369,42 +442,118 @@ def prepare_helpers
 
-      # Copies all test suite files into the suites directory in the sandbox.
+      # Creates a PowerShell hashtable from a ruby map.
+      # The only types supported for now are hash, string and Boolean.
```
nit: Missing array type

glennsarti commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -369,42 +442,118 @@ def prepare_helpers
+          "#{pad(depth)}@(\n#{array_string}\n)"
+        else
+          # When the object is not a string nor a hash or array, it will be quoted as a string.
+          %{"#{obj}"}
```
This could be dangerous if `obj` contains double quotes, as it won't be escaped.

glennsarti replied:
actually the `%{}` should handle that.

However `"#{obj}"` isn't needed and should just be `obj.to_s`

vexx32 replied:
It can also probably use an explicit comment here about how we're _relatively_ safe asusming all scalar values are strings, since PS will generally convert anything it needs to back into whatever type is has to be (especially for passing parameters).

Just in case it bites someone later on, would be good to call it out so they know what's going on. Easy enough to work around should it arise, but a comment will save some unneeded confusion.

glennsarti commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -369,42 +442,118 @@ def prepare_helpers
+        nested_directory_content = Dir.glob(File.join(path, "*/**/*"))
+        all_directory_content = [base_directory_content, nested_directory_content].flatten
+
+        all_directory_content
```
no need for a temp variable here. Just emit the line above

glennsarti commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -369,42 +442,118 @@ def prepare_helpers
+      #
+      # @api private
+      def copy_if_dir_exists(src_to_validate, destination)
+        if Dir.exist?(src_to_validate)
```
Use a guard clause instead.

```
unless Dir.exist?(src_to_validate)
  info("The modules path #{src_to_validate} was not found. Not moving to #{destination}.")
  return
end
... <do the rest of the code>
```

glennsarti commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -369,42 +442,118 @@ def prepare_helpers
+      #
+      # @api private
       def test_folder
         return config[:test_base_path] if config[:test_folder].nil?
```
Can use a ternary here

`config[:test_folder].nil? ? config[:test_base_path] : absolute_test_folder`

glennsarti commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -413,6 +562,14 @@ def absolute_test_folder
         integration_path
```
no need for a temp variable here.  Just emit the line above

gaelcolas replied:
Do you mean:
```ruby
      def absolute_test_folder
        path = (Pathname.new config[:test_folder]).realpath
        integration_path = File.join(path, "integration")
        Dir.exist?(integration_path) ? integration_path : path
      end
```

glennsarti commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -413,6 +562,14 @@ def absolute_test_folder
+      def pad(depth = 0)
+        " " * depth
+      end
+
```
Rubocop won't like the spare line :-)

glennsarti commented on spec/pester/pester_spec.rb:
```diff
@@ -14,13 +14,13 @@ def suite_test_folder
-    pester = MockPester.new
-    pester.sandboxify_path("C:/LOWERcasedpath/Pester/tests/test")
-  end
+# describe "when sandboxifying a path" do
```
Probably shouldn't have commented code here.

gaelcolas replied:
This test is failing (because of the re-write), hence why I removed it.
As it was the only test, I wanted to leave it here for future reference.

I'll probably go back and start adding some tests at some point, and that would be a good start/reference.

gaelcolas commented:
Thanks a lot @glennsarti for the review, will fix that shortly.

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -75,11 +98,38 @@ def create_sandbox
       # @return [String] a command string
       def install_command
-        return if local_suite_files.empty?
-        return if config[:use_local_pester_module]
+        # the sandbox has not yet been copied to the SUT.
+        install_command_string = <<-PS1
```
Should we rename these? These aren't really being used to install anything at the moment.

gaelcolas replied:
nope, they are method overrides from the verifier parent class.

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -75,11 +98,38 @@ def create_sandbox
+        PS1
+        really_wrap_shell_code(Util.outdent!(install_command_string))
       end
+      # PowerShellGet & Pester Bootstrap are done in prepare_command (after sandbox is transferred)
```
This comment should probably be at the top of the method.

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -97,16 +147,17 @@ def init_command
       #
       # @return [String] a command string
       def run_command
-        return if local_suite_files.empty?
```
Do we want to run even if there are no test suite files to run?

gaelcolas replied:
Yes IMO. I want to make sure Pester runs, even if there's no test to execute. At least I know my config is sound, and I just did not have any test. Otherwise the user my doubt Test-Kitchen or Kitchen-Pester...

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -148,148 +199,173 @@ def run_command_script
+              $#{module_name} = #{ps_hash(powershell_module)}
+              Write-host -noNewline 'Instaling #{module_name}'
+              Install-Module @#{module_name}
+              Write-host '... done.'
+            PS1
+          else
+            <<-PS1
+              Write-host -noNewline 'Installing #{powershell_module} ...'
+              Install-Module -Name '#{powershell_module}'
+              Write-host '... done.'
```
Nit: casing of Write-Host & parameter name

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -148,148 +199,173 @@ def run_command_script
-        <<-EOH
+        <<-PS1
           try {
+            if (!$isLinux) {
```
```suggestion
            if (!$isLinux -and !$IsMacOS) {
```

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -369,42 +442,118 @@ def prepare_helpers
+      # mostly used to copy to the sandbox.
+      #
+      # @api private
+      def copy_if_dir_exists(src_to_validate, destination)
```
I'd recommend renaming this to `copy_if_src_exists` as the name alone doesn't make it clear if it's validating the source or destination directories for existence.

And/or expand on the comment decorating it.

vexx32 commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -369,42 +442,118 @@ def prepare_helpers
+          # folder_to_create = File.basename(src_to_validate)
+          FileUtils.cp_r(src_to_validate, destination, preserve: true)
+        else
+          info("The modules path #{src_to_validate} was not found. Not moving to #{destination}.")
```
```suggestion
          info("The path #{src_to_validate} was not found. Not copying to #{destination}.")
```

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,156 @@
+[Net.ServicePointManager]::SecurityProtocol = [Net.ServicePointManager]::SecurityProtocol -bor [Net.SecurityProtocolType]::Tls12
+
+function Install-ModuleFromNuget {
+    param (
```
Nit: cmdletbinding

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,156 @@
+        $Module,
+
+        [string]
+        $GalleryUrl         = 'https://www.powershellgallery.com/api/v2'
```
```suggestion
        $GalleryUrl = 'https://www.powershellgallery.com/api/v2'
```

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,156 @@
+        $GalleryUrl         = 'https://www.powershellgallery.com/api/v2'
+    )
+
+    $tempPath       = [io.path]::GetTempPath()
```
```suggestion
    $tempPath       = [System.IO.Path]::GetTempPath()
```

nit; casing & namespace qualification

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,156 @@
+
+    Write-Verbose -Message "Downloading Package from $nupkgUrl"
+    try {
+        if (test-path $downloadedZip) {
```
```suggestion
        if (Test-Path $downloadedZip) {
```

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,156 @@
+    }
+    
+    if (-not (Test-Path -Path $downloadedZip)) {
+        Throw "Error trying to download nupkg '$nupkgUrl' to '$downloadedZip'."
```
```suggestion
        throw "Error trying to download nupkg '$nupkgUrl' to '$downloadedZip'."
```

vexx32 commented on lib/support/modules/PesterUtil/PesterUtil.psm1:
```diff
@@ -0,0 +1,156 @@
+        Write-Host "Nupkg installed at $ModuleFolder"
+    }
+
+    [GC]::Collect()
```
Should probably have this in a try/finally, wouldn't wanna leave COM stuff around if an error gets thrown.

Does that COM object have a Dispose() method we should be using?

gaelcolas replied:
The whole process is terminated shortly after bootstrap, and that's almost exclusively for WMF4.
That's also tested code I used for a while, I'd rather not potentially introduce more complexity.

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -1,5 +1,3 @@
-# -*- encoding: utf-8 -*-
-#
```
Yeah they started flagging by rubycop when I "fixed" the build...

gaelcolas commented on lib/kitchen/verifier/pester.rb:
```diff
@@ -64,9 +80,16 @@ def initialize(config = {})
+
+        debug("\n\n")
+        debug("Sandbox content:\n")
+        list_files(sandbox_path).each do |f|
+          debug("    #{f}")
+        end
```
Not I know of, but that might be in a parent object similar to `isWindows?`
I'll leave that for future improvement though, I want to get this released soon.

gaelcolas commented:
Mostly done except:
- waiting for @glennsarti to see if I understood [his comment correctly](https://github.com/test-kitchen/kitchen-pester/pull/64#discussion_r460522597)
- Will now update the help and triple check the behaviour.

I'm also [working on a few demos](https://github.com/gaelcolas/kitchenSamples) and will record a quick video.

gaelcolas commented on README.md:
```diff
@@ -20,18 +20,176 @@ gem 'kitchen-pester'
+* `downloads`- map[string, string], defaults to `["./PesterTestResults.xml"] => "./testresults"`. 
+
+
+## Options explained
```
I'll remove for simplicity then.

gaelcolas commented on README.md:
```diff
@@ -20,18 +20,176 @@ gem 'kitchen-pester'
+* `install_modules` - array, default is `[]`
+* `copy_folders` - array, default is `[]`
+* `sudo` - bool, default is `true`. (non-windows only)
+* `downloads`- map[string, string], defaults to `["./PesterTestResults.xml"] => "./testresults"`. 
```
You're correct, I just kept what was there before.

gaelcolas commented on README.md:
```diff
@@ -20,18 +20,176 @@ gem 'kitchen-pester'
+  pester_install:
+    MaximumVersion: '4.99.999'
+  bootstrap: # installs modules from nuget feed by download and unzip.
+        repository_url: "https://www.powershellgallery.com/api/v2"
+        modules:
+          - PackageManagement
+          - PowerShellGet
```
nope, it implies it's still under the `verifier:` key, but omits other elements.

vexx32 replied:
Hm? `repository_url` and `modules` are three additional indentation levels from `bootstrap` instead of one. Is that intentional? What am I missing? 😄

gaelcolas replied:
oooops, I thought you where referencing another line... fixing now... sorry.

gaelcolas commented on README.md:
````diff
@@ -20,18 +20,176 @@ gem 'kitchen-pester'
+  ```
+
+* `use_local_pester_module` - bool default is `false`  
+Skip installing pester and just use what's available on the box, or what you have copied with the `copy_folders` options.
````
nope, that option wasn't changed.

gaelcolas replied:
Afraid that if I rename to `use_installed_pester_module` people will assume you have to install it first with the other options.
`use_available_pester_module` may confuse people even more...
Suggestion?

vexx32 replied:
Tbh I'm half inclined to just say we should call it `skip_pester_install` to be clear about what it actually does; the current name is mainly just confusing.

We can change this later, though, so it's fine to leave for now. 🙂

gaelcolas replied:
better not introduce breaking changes later. I've renamed it to your suggestion, I find it clearer too.

gaelcolas commented on README.md:
```diff
@@ -20,18 +20,176 @@ gem 'kitchen-pester'
+* `test_folder` - string, default is `./tests/integration/`.
+* `remove_builtin_powershellget` - bool, default is `true`
+* `remove_builtin_pester` - bool, default is `true`
+* `bootstrap` - map,  default is `{}` (PowershellGet & Package Management)
```
removed this section in favor of the options detailed.
Updated the detailed description.

vexx32 commented:
@gaelcolas everything looks good to me, just want to clarify the one thing with indentation; I might just be missing something, just want to make sure that's right and then we should be good to merge. Thanks for all your hard work on this! 😍 

Might be worth opening an issue to track adding/updating tests as well, I suppose I'll do that when we merge it. 🙂

gaelcolas commented:
Hi,
You were right about indentation, I was not looking at the right place. fixed.

I also fixed a bug I had introduced.

I've tested PS5 and it's currently writing an error when doing the `New-PesterOption` (which is the old code I haven't changed, so it's not a regression).
I'll probably fix that in another PR once this is released.

I've written a few other samples with different OSes and it works great: https://github.com/gaelcolas/kitchenSamples

Let me know if anything else is needed.
